File: src/client/acontext-py/src/acontext/resources/sandboxes.py
"""
Sandboxes endpoints.
"""

from ..client_types import RequesterProtocol
from ..types.sandbox import (
    SandboxCommandOutput,
    SandboxRuntimeInfo,
)
from ..types.tool import FlagResponse


class SandboxesAPI:
    def __init__(self, requester: RequesterProtocol) -> None:
        self._requester = requester

    def create(self) -> SandboxRuntimeInfo:
        """Create and start a new sandbox.

        Returns:
            SandboxRuntimeInfo containing the sandbox ID, status, and timestamps.
        """
        data = self._requester.request("POST", "/sandbox")
        return SandboxRuntimeInfo.model_validate(data)

    def exec_command(
        self,
        *,
        sandbox_id: str,
        command: str,
    ) -> SandboxCommandOutput:
        """Execute a shell command in the sandbox.

        Args:
            sandbox_id: The UUID of the sandbox.
            command: The shell command to execute.

        Returns:
            SandboxCommandOutput containing stdout, stderr, and exit code.
        """
        data = self._requester.request(
            "POST",
            f"/sandbox/{sandbox_id}/exec",
            json_data={"command": command},
        )
        return SandboxCommandOutput.model_validate(data)

    def kill(self, sandbox_id: str) -> FlagResponse:
        """Kill a running sandbox.

        Args:
            sandbox_id: The UUID of the sandbox to kill.

        Returns:
            FlagResponse with status and error message.
        """
        data = self._requester.request("DELETE", f"/sandbox/{sandbox_id}")
        return FlagResponse.model_validate(data)

File: src/client/acontext-ts/src/resources/sandboxes.ts
/**
 * Sandboxes endpoints.
 */

import { RequesterProtocol } from '../client-types';
import { buildParams } from '../utils';
import {
  FlagResponse,
  FlagResponseSchema,
  GetSandboxLogsOutput,
  GetSandboxLogsOutputSchema,
  SandboxCommandOutput,
  SandboxCommandOutputSchema,
  SandboxRuntimeInfo,
  SandboxRuntimeInfoSchema,
} from '../types';

export class SandboxesAPI {
  constructor(private requester: RequesterProtocol) {}

  /**
   * Create and start a new sandbox.
   *
   * @returns SandboxRuntimeInfo containing the sandbox ID, status, and timestamps
   */
  async create(): Promise<SandboxRuntimeInfo> {
    const data = await this.requester.request('POST', '/sandbox');
    return SandboxRuntimeInfoSchema.parse(data);
  }

  /**
   * Execute a shell command in the sandbox.
   *
   * @param options - Command execution options
   * @param options.sandboxId - The UUID of the sandbox
   * @param options.command - The shell command to execute
   * @param options.timeout - Optional timeout in milliseconds for this command.
   *                          If not provided, uses the client's default timeout.
   * @returns SandboxCommandOutput containing stdout, stderr, and exit code
   */
  async execCommand(options: {
    sandboxId: string;
    command: string;
    timeout?: number;
  }): Promise<SandboxCommandOutput> {
    const data = await this.requester.request(
      'POST',
      `/sandbox/${options.sandboxId}/exec`,
      {
        jsonData: { command: options.command },
        timeout: options.timeout,
      }
    );
    return SandboxCommandOutputSchema.parse(data);
  }

  /**
   * Kill a running sandbox.
   *
   * @param sandboxId - The UUID of the sandbox to kill
   * @returns FlagResponse with status and error message
   */
  async kill(sandboxId: string): Promise<FlagResponse> {
    const data = await this.requester.request(
      'DELETE',
      `/sandbox/${sandboxId}`
    );
    return FlagResponseSchema.parse(data);
  }

  /**
   * Get sandbox logs for the project with cursor-based pagination.
   *
   * @param options - Optional parameters for retrieving logs
   * @param options.limit - Maximum number of logs to return (default 20, max 200)
   * @param options.cursor - Cursor for pagination. Use the cursor from the previous response to get the next page
   * @param options.timeDesc - Order by created_at descending if true, ascending if false (default false)
   * @returns GetSandboxLogsOutput containing the list of sandbox logs and pagination information
   */
  async getLogs(options?: {
    limit?: number | null;
    cursor?: string | null;
    timeDesc?: boolean | null;
  }): Promise<GetSandboxLogsOutput> {
    const params = buildParams({
      limit: options?.limit ?? null,
      cursor: options?.cursor ?? null,
      time_desc: options?.timeDesc ?? null,
    });
    const data = await this.requester.request('GET', '/sandbox/logs', {
      params: Object.keys(params).length > 0 ? params : undefined,
    });
    return GetSandboxLogsOutputSchema.parse(data);
  }
}
